alf/utils/tensorrt_utils.py

```diff
@@ -0,0 +1,390 @@
+        2. Convert tensor inputs to numpy inputs for the inference engine.
+        3. Convert the outputs of the engine to tensor outputs.
+        4. Call ``_OnnxWrapper.recover_module_output`` to recover the original
+           structure of the output.
+
+        Args:
+            module: The module to be converted.
+            method: A method in ``module`` to be converted.
+            onnx_file: The path to the onnx model. If None, no external file will
+                be created. Instead, the onnx model will be created in memory.
+            onnx_verbose: If True, the onnx model exporting process will output
+                verbose information.
+            device: The device used by TensorRT to run the onnx model. Default
+                is 'CUDA:0'. (CPU is not allowed.)
+            example_args: The example args to be used for ``method``. Should be
+                a tuple of args.
+            example_kwargs: The example kwargs to be used for ``method``. Should
+                be a dict of kwargs.
+        """
+        super().__init__()
+        self._onnx_wrapper = _OnnxWrapper(module, method, example_args,
+                                          example_kwargs)
+
+        flat_all_args = alf.nest.flatten([example_args, example_kwargs])
+        flat_all_args = self._dtype_conversions(flat_all_args)
+
+        if onnx_file is None:
+            onnx_io = io.BytesIO()
+        else:
+            onnx_io = onnx_file
+        # 'args' must be a tuple of tensors
+        torch.onnx.export(
+            self._onnx_wrapper,
+            args=flat_all_args,
+            f=onnx_io,
+            # Don't modify the version easily! Other versions might
+            # have weird errors.
+            opset_version=12,
+            verbose=onnx_verbose)
+        if isinstance(onnx_io, io.BytesIO):
+            onnx_io.seek(0)
+        onnx_model = onnx.load(onnx_io)
+
+        # Infer shapes first to avoid the error: "Please run shape inference on the onnx model first."
+        # from TensorRT
+        onnx_model = shape_inference.infer_shapes(onnx_model)
+        self._engine = backend.prepare(onnx_model, device=device.upper())
+
+    def _dtype_conversions(
+            self, flat_all_args: Tuple[torch.Tensor]) -> Tuple[torch.Tensor]:
+        # Convert all uint8 tensors to int32, float64 to float32, because tensorRT
+        # does not support uint8 or float64 tensors
+        dtype_mappings = {
+            torch.uint8: torch.int32,
+            torch.float64: torch.float32
+        }
+
+        def _convert_if_needed(x):
+            if x.dtype in dtype_mappings:
+                return x.to(dtype=dtype_mappings[x.dtype])
+            return x
+
+        return tuple([_convert_if_needed(a) for a in flat_all_args])
+
+    def forward(self, *args, **kwargs):
+        flat_all_args = self._dtype_conversions(
+            alf.nest.flatten([args, kwargs]))
+        flat_all_args_np = [x.detach().cpu().numpy() for x in flat_all_args]
```

Comment:
The arguments and outputs are always numpy arrays?

Reply:
> The arguments and outputs are always numpy arrays?

Right. ONNX or tensorRT doesn't assume any universal data type for gpu tensors.